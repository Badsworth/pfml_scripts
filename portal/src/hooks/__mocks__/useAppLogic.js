import Claim from "../../models/Claim";
import ClaimCollection from "../../models/ClaimCollection";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  auth: {
    createAccount: jest.fn(),
    forgotPassword: jest.fn(),
    login: jest.fn(),
  },
  claims: new ClaimCollection(),
  createClaim: jest.fn(() => new Claim({ application_id: uniqueId() })),
  loadClaims: jest.fn(),
  submitClaim: jest.fn(),
  updateClaim: jest.fn(),
}));
