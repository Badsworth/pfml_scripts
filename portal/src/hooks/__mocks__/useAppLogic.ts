import {
  BenefitsApplicationDocument,
  ClaimDocument,
} from "../../models/Document";
import User, { UserLeaveAdministrator } from "../../models/User";
import ApiResourceCollection from "../../models/ApiResourceCollection";
import BenefitsApplication from "../../models/BenefitsApplication";
import Claim from "../../models/Claim";
import ClaimDetail from "../../models/ClaimDetail";
import EmployerClaim from "../../models/EmployerClaim";
import Flag from "../../models/Flag";
import { Payment } from "src/models/Payment";
import { uniqueId } from "lodash";

export default jest.fn(() => ({
  errors: [],
  auth: {
    createAccount: jest.fn(),
    createEmployerAccount: jest.fn(),
    forgotPassword: jest.fn(),
    isLoggedIn: true,
    login: jest.fn(),
    logout: jest.fn(),
    requireLogin: jest.fn(),
    resendForgotPasswordCode: jest.fn(),
    resendVerifyAccountCode: jest.fn(),
    resetPassword: jest.fn(),
    verifyAccount: jest.fn(),
  },
  catchError: jest.fn(),
  benefitsApplications: {
    benefitsApplications: new ApiResourceCollection<BenefitsApplication>(
      "application_id"
    ),
    complete: jest.fn(),
    create: jest.fn(
      () => new BenefitsApplication({ application_id: uniqueId() })
    ),
    get: jest.fn(),
    hasLoadedBenefitsApplicationAndWarnings: jest.fn().mockReturnValue(true),
    hasLoadedAll: false,
    load: jest.fn(),
    loadAll: jest.fn(),
    submit: jest.fn(),
    submitPaymentPreference: jest.fn(),
    update: jest.fn(),
    warningsLists: {},
  },
  claims: {
    claimDetail: new ClaimDetail(),
    claims: new ApiResourceCollection<Claim>("fineos_absence_id"),
    clearClaims: jest.fn(),
    isLoadingClaims: null,
    isLoadingClaimDetail: null,
    loadClaimDetail: jest.fn(),
    paginationMeta: {},
    loadPage: jest.fn(),
  },
  clearErrors: jest.fn(),
  clearRequiredFieldErrors: jest.fn(),
  documents: {
    attach: jest.fn((application_id, files) => {
      const uploadPromises: Array<Promise<{ success: boolean }>> = [];
      for (let i = 0; i < files.length; i++) {
        uploadPromises.push(Promise.resolve({ success: true }));
      }
      return uploadPromises;
    }),
    hasLoadedClaimDocuments: jest.fn(),
    documents: new ApiResourceCollection<BenefitsApplicationDocument>(
      "fineos_document_id"
    ),
    download: jest.fn(),
    loadAll: jest.fn(),
  },
  employers: {
    addEmployer: jest.fn(),
    downloadDocument: jest.fn(() => new Blob()),
    loadClaim: jest.fn(
      () =>
        new EmployerClaim({
          absence_periods: [],
          fineos_absence_id: "NTN-111-ABS-01",
        })
    ),
    loadDocuments: jest.fn(
      () => new ApiResourceCollection<ClaimDocument>("fineos_document_id")
    ),
    loadWithholding: jest.fn(() => ({ filing_period: "2011-11-20" })),
    submitClaimReview: jest.fn(),
    submitWithholding: jest.fn(),
  },
  payments: {
    loadPayments: jest.fn(),
    loadedPaymentsData: new Payment(),
    hasLoadedPayments: true,
    isLoadingPayments: false,
  },
  portalFlow: {
    getNextPageRoute: jest.fn(),
    goTo: jest.fn(),
    goToNextPage: jest.fn(),
    goToPageFor: jest.fn(),
    pathname: "/mock-portalFlow-pathname",
  },
  otherLeaves: {
    removeEmployerBenefit: jest.fn(() => true),
    removeOtherIncome: jest.fn(() => true),
    removePreviousLeave: jest.fn(() => true),
  },
  setErrors: jest.fn(),
  updateUser: jest.fn(),
  user: new User({
    user_id: "mock_user_id",
    consented_to_data_sharing: true,
    email_address: "ali@miau.com",
    auth_id: "cognito_123",
    roles: [],
    user_leave_administrators: [],
  }),
  users: {
    loadUser: jest.fn(),
    requireUserConsentToDataAgreement: jest.fn(),
    requireUserRole: jest.fn(),
    updateUser: jest.fn(),
    user: new User({
      auth_id: "mock_auth_id",
      user_id: "mock_user_id",
      consented_to_data_sharing: true,
      user_leave_administrators: [
        // not verified, but can be
        new UserLeaveAdministrator({
          employer_dba: "Book Bindings 'R Us",
          employer_fein: "**-***1823",
          employer_id: "dda903f-f093f-ff900",
          has_fineos_registration: true,
          has_verification_data: true,
          verified: false,
        }),
        // already verified
        new UserLeaveAdministrator({
          employer_dba: "Knitting Castle",
          employer_fein: "**-***3443",
          employer_id: "dda930f-93jfk-iej08",
          has_fineos_registration: true,
          has_verification_data: true,
          verified: true,
        }),
        // not verified and cannot be verified
        new UserLeaveAdministrator({
          employer_dba: "Tomato Touchdown",
          employer_fein: "**-***7192",
          employer_id: "io19fj9-00jjf-uiw3r",
          has_fineos_registration: true,
          has_verification_data: false,
          verified: false,
        }),
      ],
    }),
  },
  featureFlags: {
    flags: [
      new Flag({
        enabled: false,
        name: "maintenance",
        options: {
          page_routes: ["/*"],
        },
        start: null,
        end: null,
      }),
    ],
    getFlag: jest.fn(() => {
      return new Flag({
        enabled: false,
        name: "maintenance",
        options: {
          page_routes: ["/*"],
        },
        start: null,
        end: null,
      });
    }),
    loadFlags: jest.fn(),
  },
}));
