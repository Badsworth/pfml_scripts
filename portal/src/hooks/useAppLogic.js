/* eslint sort-keys: ["error", "asc"] */
import useAppErrorsLogic from "./useAppErrorsLogic";
import useAuthLogic from "./useAuthLogic";
import useBenefitsApplicationsLogic from "./useBenefitsApplicationsLogic";
import useClaimsLogic from "./useClaimsLogic";
import useDocumentsLogic from "./useDocumentsLogic";
import useEmployersLogic from "./useEmployersLogic";
import useFlagsLogic from "./useFeatureFlagsLogic";
import useOtherLeavesLogic from "./useOtherLeavesLogic";
import usePortalFlow from "./usePortalFlow";
import useUsersLogic from "./useUsersLogic";

const useAppLogic = () => {
  // Utility to determine application flow
  const portalFlow = usePortalFlow();

  // State representing currently visible errors and warnings
  const { appErrors, ...appErrorsLogic } = useAppErrorsLogic({ portalFlow });
  const auth = useAuthLogic({ appErrorsLogic, portalFlow });

  // State representing the Portal's user object.
  // Initialize to empty user but will be populated upon the first API call
  // to fetch the user (or create the user on their first login)
  const users = useUsersLogic({
    appErrorsLogic,
    isLoggedIn: auth.isLoggedIn,
    portalFlow,
  });

  // user will be eventually set through a `login` method
  // const [user, setUser] = useState();
  // TODO (CP-885): remove user from configs and write login method
  const benefitsApplications = useBenefitsApplicationsLogic({
    appErrorsLogic,
    portalFlow,
    user: users.user,
  });

  const claims = useClaimsLogic({
    appErrorsLogic,
  });

  const documents = useDocumentsLogic({
    appErrorsLogic,
    portalFlow,
  });

  const employers = useEmployersLogic({
    appErrorsLogic,
    portalFlow,
    setUser: users.setUser,
  });

  const otherLeaves = useOtherLeavesLogic({ appErrorsLogic });
  const featureFlags = useFlagsLogic({ appErrorsLogic });

  return {
    appErrors,
    auth,
    benefitsApplications,
    catchError: appErrorsLogic.catchError,
    claims,
    // TODO (CP-886): remove once all API calls are behind appLogic
    clearErrors: appErrorsLogic.clearErrors,
    documents,
    employers,
    otherLeaves,
    portalFlow,
    setAppErrors: appErrorsLogic.setAppErrors,
    users,
    featureFlags,
  };
};

export default useAppLogic;
