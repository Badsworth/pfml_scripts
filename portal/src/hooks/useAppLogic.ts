/* eslint sort-keys: ["error", "asc"] */
import useApplicationImportsLogic from "./useApplicationImportsLogic";
import useAuthLogic from "./useAuthLogic";
import useBenefitYearsLogic from "./useBenefitYearsLogic";
import useBenefitsApplicationsLogic from "./useBenefitsApplicationsLogic";
import useClaimsLogic from "./useClaimsLogic";
import useDocumentsLogic from "./useDocumentsLogic";
import useEmployersLogic from "./useEmployersLogic";
import useErrorsLogic from "./useErrorsLogic";
import useFeatureFlagsLogic from "./useFeatureFlagsLogic";
import useHolidaysLogic from "./useHolidaysLogic";
import usePaymentsLogic from "./usePaymentsLogic";
import usePortalFlow from "./usePortalFlow";
import useUsersLogic from "./useUsersLogic";

const useAppLogic = () => {
  // Utility to determine application flow
  const portalFlow = usePortalFlow();

  // State representing currently visible errors and warnings
  const errorsLogic = useErrorsLogic({ portalFlow });

  const applicationImports = useApplicationImportsLogic({
    errorsLogic,
    portalFlow,
  });
  const auth = useAuthLogic({ errorsLogic, portalFlow });

  // State representing the Portal's user object.
  // Initialize to empty user but will be populated upon the first API call
  // to fetch the user (or create the user on their first login)
  const users = useUsersLogic({
    errorsLogic,
    isLoggedIn: !!auth.isLoggedIn,
    portalFlow,
  });

  const benefitYears = useBenefitYearsLogic({ errorsLogic });

  const benefitsApplications = useBenefitsApplicationsLogic({
    errorsLogic,
    portalFlow,
  });

  const claims = useClaimsLogic({
    errorsLogic,
    portalFlow,
  });

  const documents = useDocumentsLogic({
    errorsLogic,
  });

  const employers = useEmployersLogic({
    clearClaims: claims.clearClaims,
    errorsLogic,
    portalFlow,
    setUser: users.setUser,
  });

  const featureFlags = useFeatureFlagsLogic();

  const holidays = useHolidaysLogic({
    errorsLogic,
  });

  const payments = usePaymentsLogic({
    errorsLogic,
  });

  return {
    // `_errorsLogic` should not be used except for testing
    _errorsLogic: errorsLogic,
    applicationImports,
    auth,
    benefitYears,
    benefitsApplications,
    catchError: errorsLogic.catchError,
    claims,
    // TODO (CP-886): remove once all API calls are behind appLogic
    clearErrors: errorsLogic.clearErrors,
    clearRequiredFieldErrors: errorsLogic.clearRequiredFieldErrors,
    documents,
    employers,
    errors: errorsLogic.errors,
    featureFlags,
    holidays,
    payments,
    portalFlow,
    setErrors: errorsLogic.setErrors,
    users,
  };
};

export default useAppLogic;
export type AppLogic = ReturnType<typeof useAppLogic>;
