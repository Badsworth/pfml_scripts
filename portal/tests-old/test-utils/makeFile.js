/**
 * Make a File instance
 * @see https://developer.mozilla.org/en-US/docs/Web/API/File
 * @param {object} attrs File attributes
 * @param {string} attrs.name The file's name
 * @param {string} attrs.type The file's MIME type
 * @returns {File}
 */
export const makeFile = (attrs = {}) => {
  const { name, type } = Object.assign(
    {
      name: "file.pdf",
      type: "application/pdf",
    },
    attrs
  );

  return new File([], name, { type });
};

/**
 * Create an HTML <input> for a test's event.target
 * @param {object} attrs - HTML attributes for the input
 * @returns {HTMLInputElement}
 */
export function createInputElement(attrs) {
  const input = document.createElement("input");

  Object.entries(attrs).forEach(([attributeName, value]) => {
    input.setAttribute(attributeName, value);
  });

  if (attrs.checked === false) input.checked = false;

  return input;
}
