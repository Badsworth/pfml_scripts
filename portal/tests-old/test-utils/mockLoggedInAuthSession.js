import { Auth } from "@aws-amplify/auth";

/**
 * Mock the Cognito authentication session so a test user appears logged in
 */
const mockLoggedInAuthSession = () => {
  const accessTokenJwt =
    "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiQnVkIn0.YDRecdsqG_plEwM0H8rK7t2z0R3XRNESJB5ZXk-FRN8";

  jest.spyOn(Auth, "currentSession").mockImplementation(() =>
    Promise.resolve({
      accessToken: { jwtToken: accessTokenJwt },
    })
  );
};

export default mockLoggedInAuthSession;
