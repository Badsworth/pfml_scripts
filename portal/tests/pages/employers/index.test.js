import Index from "../../../src/pages/employers";
import React from "react";
import { shallow } from "enzyme";
import { testHook } from "../../test-utils";
import useAppLogic from "../../../src/hooks/useAppLogic";

jest.mock("../../../src/hooks/useAppLogic");

describe("Employer index", () => {
  let appLogic, wrapper;

  beforeEach(() => {
    testHook(() => {
      appLogic = useAppLogic();
    });

    wrapper = shallow(<Index appLogic={appLogic} />);
  });

  it("renders pre-launch content when claimantShowMedicalLeaveType is false", () => {
    process.env.featureFlags = {
      claimantShowMedicalLeaveType: false,
    };

    expect(wrapper).toMatchSnapshot();
  });

  it("renders post-launch content when claimantShowMedicalLeaveType is true", () => {
    process.env.featureFlags = {
      claimantShowMedicalLeaveType: true,
    };

    expect(wrapper).toMatchSnapshot();
  });
});
