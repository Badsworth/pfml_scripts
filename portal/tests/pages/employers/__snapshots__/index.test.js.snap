// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employer index renders post-launch content when claimantShowMedicalLeaveType is true 1`] = `
<div
  className="grid-container"
>
  <div
    className="grid-row"
  >
    <div
      className="desktop:grid-col-8"
    >
      <Title>
        Welcome
      </Title>
      <p>
        Thanks for joining the paid leave program. Massachusetts workers can now apply for paid family leave for bonding with a child, with more benefits available in 2021.
      </p>
      <Heading
        level="2"
      >
        <SvgIconMail
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        Check your email regularly
      </Heading>
      <p>
        When an employee applies for leave, you’ll receive email updates about their application status and any steps you need to take. We’ll include everyone who has an employer account with your company in case you’re out of the office.
      </p>
      <Heading
        level="2"
      >
        <IconWait
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        Respond to applications within 10 business days
      </Heading>
      <p>
        When an application is submitted, you have 10 business days to open the direct link from your email and review it online. You can comment on the application, approve or deny it, and report fraud if needed. Reviewing takes about 10 minutes. If we don’t hear from anyone at your company before the deadline, we’ll process the application solely based on the information the employee provided.
      </p>
      <Heading
        level="2"
      >
        <SvgIconPdf
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        View forms and notices online
      </Heading>
      <p>
        <Trans
          components={
            Object {
              "healthcare-provider-form-link": <a
                href="https://www.mass.gov/hcp-form"
                rel="noopener"
                target="_blank"
              />,
            }
          }
          i18nKey="pages.employersDashboard.viewFormsBody"
        />
      </p>
    </div>
    <div
      className="grid-col-fill"
    />
    <aside
      className="desktop:grid-col-3 margin-top-7 desktop:margin-top-1"
    >
      <Heading
        level="2"
      >
        Learn more
      </Heading>
      <Trans
        components={
          Object {
            "li": <li />,
            "mass-employer-role-link": <a
              href="https://www.mass.gov/info-details/employer-paid-leave-application-review-guide"
              rel="noopener"
              target="_blank"
            />,
            "reimbursements-link": <a
              href="https://www.mass.gov/forms/request-reimbursement-for-your-employees-paid-leave-benefits"
              rel="noopener"
              target="_blank"
            />,
            "ul": <ul
              className="usa-list desktop:font-body-2xs desktop:padding-top-05"
            />,
          }
        }
        i18nKey="pages.employersDashboard.learnMoreLinks"
      />
    </aside>
  </div>
</div>
`;

exports[`Employer index renders post-launch content when claimantShowMedicalLeaveType is true 2`] = `
Array [
  "You’ll get an email about our application decision with a direct link to download the letter your employee received. For medical leave, you can download the ",
  <a
    href="https://www.mass.gov/hcp-form"
    key="1"
    rel="noopener"
    target="_blank"
  >
    Certification of a Serious Health Condition form
  </a>,
  " during the review process.",
]
`;

exports[`Employer index renders post-launch content when claimantShowMedicalLeaveType is true 3`] = `
<ul
  className="usa-list desktop:font-body-2xs desktop:padding-top-05"
  key="0"
>
  <li
    key="0"
  >
    <a
      href="https://www.mass.gov/info-details/employer-paid-leave-application-review-guide"
      key="0"
      rel="noopener"
      target="_blank"
    >
      Your role as a Massachusetts employer
    </a>
  </li>
  <li
    key="1"
  >
    <a
      href="https://www.mass.gov/forms/request-reimbursement-for-your-employees-paid-leave-benefits"
      key="0"
      rel="noopener"
      target="_blank"
    >
      Employer reimbursements
    </a>
  </li>
</ul>
`;

exports[`Employer index renders pre-launch content when claimantShowMedicalLeaveType is false 1`] = `
<div
  className="grid-container"
>
  <div
    className="grid-row"
  >
    <div
      className="desktop:grid-col-8"
    >
      <Title>
        Welcome
      </Title>
      <p>
        Thanks for joining the paid leave program. Massachusetts workers can now apply for paid family leave for bonding with a child, with more benefits available in 2021.
      </p>
      <Heading
        level="2"
      >
        <SvgIconMail
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        Check your email regularly
      </Heading>
      <p>
        When an employee applies for leave, you’ll receive email updates about their application status and any steps you need to take. We’ll include everyone who has an employer account with your company in case you’re out of the office.
      </p>
      <Heading
        level="2"
      >
        <IconWait
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        Respond to applications within 10 business days
      </Heading>
      <p>
        When an application is submitted, you have 10 business days to open the direct link from your email and review it online. You can comment on the application, approve or deny it, and report fraud if needed. Reviewing takes about 10 minutes. If we don’t hear from anyone at your company before the deadline, we’ll process the application solely based on the information the employee provided.
      </p>
      <Heading
        level="2"
      >
        <SvgIconPdf
          className="margin-right-2 text-secondary text-middle"
          fill="currentColor"
          height={30}
          width={30}
        />
        View forms and notices online
      </Heading>
      <p>
        <Trans
          components={
            Object {
              "healthcare-provider-form-link": <a
                href="https://www.mass.gov/hcp-form"
                rel="noopener"
                target="_blank"
              />,
            }
          }
          i18nKey="pages.employersDashboard.viewFormsBody"
        />
      </p>
    </div>
    <div
      className="grid-col-fill"
    />
    <aside
      className="desktop:grid-col-3 margin-top-7 desktop:margin-top-1"
    >
      <Heading
        level="2"
      >
        Learn more
      </Heading>
      <Trans
        components={
          Object {
            "li": <li />,
            "mass-employer-role-link": <a
              href="https://www.mass.gov/info-details/employer-paid-leave-application-review-guide"
              rel="noopener"
              target="_blank"
            />,
            "reimbursements-link": <a
              href="https://www.mass.gov/forms/request-reimbursement-for-your-employees-paid-leave-benefits"
              rel="noopener"
              target="_blank"
            />,
            "ul": <ul
              className="usa-list desktop:font-body-2xs desktop:padding-top-05"
            />,
          }
        }
        i18nKey="pages.employersDashboard.learnMoreLinks"
      />
    </aside>
  </div>
</div>
`;

exports[`Employer index renders pre-launch content when claimantShowMedicalLeaveType is false 2`] = `
Array [
  "You’ll get an email about our application decision with a direct link to download the letter your employee received. For medical leave, you can download the ",
  <a
    href="https://www.mass.gov/hcp-form"
    key="1"
    rel="noopener"
    target="_blank"
  >
    Certification of a Serious Health Condition form
  </a>,
  " during the review process.",
]
`;

exports[`Employer index renders pre-launch content when claimantShowMedicalLeaveType is false 3`] = `
<ul
  className="usa-list desktop:font-body-2xs desktop:padding-top-05"
  key="0"
>
  <li
    key="0"
  >
    <a
      href="https://www.mass.gov/info-details/employer-paid-leave-application-review-guide"
      key="0"
      rel="noopener"
      target="_blank"
    >
      Your role as a Massachusetts employer
    </a>
  </li>
  <li
    key="1"
  >
    <a
      href="https://www.mass.gov/forms/request-reimbursement-for-your-employees-paid-leave-benefits"
      key="0"
      rel="noopener"
      target="_blank"
    >
      Employer reimbursements
    </a>
  </li>
</ul>
`;
